flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,270 @@
+
+            return new TaskBalancedPreferredSlotSharingStrategy(
+                    topology, logicalSlotSharingGroups, coLocationGroups);
+        }
+    }
+
+    /** SlotSharingGroupBuilder class for balanced scheduling strategy. */
+    private static class TaskBalancedExecutionSlotSharingGroupBuilder {
+
+        private final SchedulingTopology topology;
+
+        private final Map<JobVertexID, SlotSharingGroup> slotSharingGroupMap;
+
+        /**
+         * Record the number of {@link ExecutionSlotSharingGroup}s for {@link SlotSharingGroup}s.
+         */
+        private final Map<SlotSharingGroup, List<ExecutionSlotSharingGroup>> ssgToExecutionSSGs;
+
+        /**
+         * Record the next round-robin {@link ExecutionSlotSharingGroup} index for {@link
+         * SlotSharingGroup}s.
+         */
+        private final Map<SlotSharingGroup, Integer> slotRoundRobinIndex;
+
+        private final Map<ExecutionVertexID, ExecutionSlotSharingGroup>
+                executionSlotSharingGroupMap;
+
+        private final Map<JobVertexID, CoLocationGroup> coLocationGroupMap;
+
+        private final Map<CoLocationConstraint, ExecutionSlotSharingGroup> clcToESsgMap;
+
+        private TaskBalancedExecutionSlotSharingGroupBuilder(
+                final SchedulingTopology topology,
+                final Set<SlotSharingGroup> logicalSlotSharingGroups,
+                final Set<CoLocationGroup> coLocationGroups) {
+            this.topology = checkNotNull(topology);
+
+            this.coLocationGroupMap = new HashMap<>();
+            for (CoLocationGroup coLocationGroup : coLocationGroups) {
+                for (JobVertexID jobVertexId : coLocationGroup.getVertexIds()) {
+                    coLocationGroupMap.put(jobVertexId, coLocationGroup);
+                }
+            }
+
+            this.clcToESsgMap = new HashMap<>();
+            this.ssgToExecutionSSGs = new HashMap<>(logicalSlotSharingGroups.size());
+            this.slotRoundRobinIndex = new HashMap<>(logicalSlotSharingGroups.size());
+            this.slotSharingGroupMap = new HashMap<>();
+            this.executionSlotSharingGroupMap = new HashMap<>();
+
+            for (SlotSharingGroup slotSharingGroup : logicalSlotSharingGroups) {
+                for (JobVertexID jobVertexId : slotSharingGroup.getJobVertexIds()) {
+                    slotSharingGroupMap.put(jobVertexId, slotSharingGroup);
+                }
+            }
+        }
+
+        private Map<ExecutionVertexID, ExecutionSlotSharingGroup> build() {
+            final LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> allVertices =
+                    getExecutionVertices(topology);
+
+            computeSsgToESsgMap(allVertices);
+
+            // Loop on job vertices
+            for (Map.Entry<JobVertexID, List<SchedulingExecutionVertex>> entry :
+                    allVertices.entrySet()) {
+                SlotSharingGroup ssg = slotSharingGroupMap.get(entry.getKey());
```

Comment:
We'd set such "entry.getKey()" to a new variable for better readablility. Same as others.